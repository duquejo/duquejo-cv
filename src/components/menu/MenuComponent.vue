<template>
  <nav class="lg:flex lg:flex-col lg:items-center lg:justify-between hidden lg:my-4 relative">
    <LangSelectComponent id="menu-language-selector" />
    <div v-if="!isLoading && events" class="fixed w-96 h-auto top-0 bottom-0 my-4 rounded z-10 bg-gray-100 shadow-md flex flex-col justify-between gap-2 transition-all duration-500 ease-in-out" :class="menuClasses">
      <h4 class="mt-8 mb-2.5 text-h4 mx-4 text-center">{{ t('general.menu.events') }}</h4>
      <div class="overflow-y-auto">
          <RecentCardComponent v-for="event in events" :key="event.id" :event="event" />
          <p class="mx-4 mt-0 text-xs text-center">
            <a href="https://github.com/duquejo01" target="_blank" :aria-label="t('general.menu.events_repo_general_link_label')">{{ t('general.menu.events_repo_general_link') }}</a>
          </p>
      </div>
    </div>
    <i v-if="!isLoading" class="icon" @click="onLoadClick">
      <v-icon :class="eventIconClass" :name="eventIcon" scale="1.5" fill="black"/>
    </i>
    <i v-else class="icon cursor-default" :title="t('general.menu.events_loading_description')">
      <v-icon name="fa-spinner" scale="1" fill="gray" animation="spin"/>
    </i>
  </nav>
</template>

<script setup lang="ts">
import { computed, defineAsyncComponent } from 'vue';
import { useI18n } from 'vue-i18n';

import useQueryClient from '@/composables/useQueryClient';

const LangSelectComponent = defineAsyncComponent(() => import('@/components/language-select/LangSelectComponent.vue'));
const RecentCardComponent = defineAsyncComponent(() => import('@/components/recent-card/RecentCardComponent.vue'));

interface Props {
  isOpen?: boolean;
}

interface Emits {
  (e: 'toggle' ): void;
}

const { t } = useI18n();
const { events, isLoading, refetch } = useQueryClient();

const props = withDefaults(defineProps<Props>(), {
  isOpen: false,
});
const emit = defineEmits<Emits>();

const menuClasses = computed(() => ({
  'opacity-0 translate-x-full right-0': ! props.isOpen,
  'opacity-100 translate-x-0 right-12': props.isOpen,
}));

const eventIcon = computed(() => props.isOpen ? 'hi-solid-light-bulb' : 'hi-light-bulb' );
const eventIconClass = computed(() => ({
  'animate-bounce': ! props.isOpen,
  'animate-pulse': props.isOpen,
}));

const onLoadClick = () => {
  refetch();
  emit('toggle');
};

</script>