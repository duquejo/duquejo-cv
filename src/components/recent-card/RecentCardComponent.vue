
<template>
  <div class="bg-white rounded pl-5 pr-4 py-4 last:mb-8 first:mt-1 mx-2 my-4 shadow flex-col">
    <time v-if="parseDate" class="text-xs font-extrabold text-right">{{ parseDate }}</time>
    <div class="flex items-center text-xs mb-2 justify-between gap-x-5">
      <span class="flex items-center gap-x-2">
        <img :src="event.actor.avatar_url" :alt="`@${event.actor.display_login}`" class="max-h-7 rounded-full" />
        <a class="font-semibold underline underline-offset-2" :href="`https://github.com/${event.repo.name}`" target="_blank">{{ event.repo.name }}</a>
      </span>
      <PillComponent 
        v-if="event.payload.ref && event.type === EventType.CreateEvent" 
        :text="event.payload.ref" 
        :title="event.payload.ref" 
        size="small"
      />
      <PillComponent 
        class="truncate"
        v-else-if="event.payload.pull_request && event.type === EventType.PullRequestEvent" 
        :text="event.payload.pull_request.head.ref" 
        :title="event.payload.pull_request.head.ref"
        size="small"
      />
    </div>
    <p v-if="! [EventType.PullRequestEvent, EventType.WatchEvent].includes(event.type)" class="bg-gray-100 rounded-lg p-3 my- text-xs leading-relaxed text-left">
      <strong>{{ event.payload.description }}</strong>
      <ul v-if="event.payload.commits" class="block">
        <li v-for="commit in event.payload.commits" :key="commit.sha">
          {{ markdownText(commit.message) }}
          <a
            v-if="event.type === 'PushEvent'"
            class="text-[0.7rem] font-semibold text-gray-700"
            target="_blank"
            :href="`${commit.url.replace('https://api.github.com/repos', 'https://github.com')}`"
            :title="t('general.menu.events_repo_commit_link')">
            <v-icon name="bi-github" scale="0.8"/>
          </a>
        </li>
      </ul>
    </p>
    <div data-test="rc-type" class="flex items-center px-2 py-2 rounded-3xl gap-x-3 shadow-md relative hover:translate-y-0.5 transition-transform ring-1 ring-gray-200 text-xs">
      <span class="p-1 rounded-full leading-none max-h-7" :class="icon.color">
        <component is="v-icon" :name="icon.icon" scale="1" />
      </span>
      {{ event.type }}
    </div>
  </div>
</template>

<script setup lang="ts">
import { toRefs, computed } from 'vue';
import { useI18n } from 'vue-i18n';
import * as emoji from 'node-emoji';

import type { Event } from '@/interfaces/github-events';
import { EventType } from '@/interfaces/github-events';
import PillComponent from '@/components/pill/PillComponent.vue';


interface Props {
  event: Event
}

const props = defineProps<Props>();
const { event } = toRefs(props);

const { t } = useI18n();

const markdownText = (text: string): string => emoji.emojify(text);

const icon = computed(() => {
  switch (event.value.type) {
    case EventType.PullRequestEvent:
      return {
        icon: 'io-git-pull-request-sharp',
        color: 'bg-yellow-400',
      };
    case EventType.CreateEvent:
      return {
        icon: 'io-git-branch',
        color: 'bg-teal-400',
      };
    case EventType.PushEvent:
      return {
        icon: 'oi-repo-push',
        color: 'bg-red-400',
      };
    case EventType.WatchEvent: 
      return {
        icon: 'ri-eye-2-line',
        color: 'bg-cyan-300',
      };
    default:
      return {
        icon: 'oi-repo-push'
      };
  };
});

const parseDate = computed(() => {
  const date = new Date(event.value.created_at);

  if( ! ( date instanceof Date && ! isNaN(date.getTime()) ) ) {
    return false
  }

  return date.toLocaleString();
});
</script>