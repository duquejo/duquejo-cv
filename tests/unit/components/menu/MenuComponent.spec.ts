import { beforeEach, describe, expect, it, vi } from 'vitest';
import { VueWrapper, shallowMount } from '@vue/test-utils';

import { useI18n } from 'vue-i18n';

import createVueI18n from '../../../mocks/mock-i18n';
import MenuComponent from '../../../../src/components/menu/MenuComponent.vue';
import RecentCardComponent from '../../../../src/components/recent-card/RecentCardComponent.vue';
import createVueQuery from '../../../mocks/mock-vueQuery';

vi.mock('vue-i18n', async () => {
  const actual = await vi.importActual('vue-i18n');
  return {
    ...actual,
    useI18n: vi.fn(),
  };
});

const mockUseI18n = vi.mocked(useI18n);

describe('+ MenuComponent.vue tests', () => {

  const i18n = createVueI18n('en');

  let wrapper: VueWrapper;

  beforeEach(() => {
    vi.clearAllMocks();
    mockUseI18n.mockReturnValue({
      t: (key: string) => key,
    });
    wrapper = shallowMount(MenuComponent, {
      global: {
        plugins: [i18n, createVueQuery],
      }
    });
  });

  it('Should match the snapshot with the defaults', () => {
    expect(wrapper.html()).toMatchSnapshot();

    expect(wrapper.findAllComponents(RecentCardComponent)).toHaveLength(0);

    const eventsIcon = wrapper.find('.icon');
    expect(eventsIcon.exists()).toBeTruthy();
    expect(eventsIcon.attributes('title')).toBe('general.menu.events_loading_description');
  });

  it('Should retrieve the translations', () => {
    expect(mockUseI18n).toHaveBeenCalled();
  });

  it('should emit the toggle event if the menu icon is triggered', async () => {
    const wrapperWithProps = shallowMount(MenuComponent, {
      props: {
        isOpen: false,
      },
      global: {
        plugins: [i18n, createVueQuery],
      }
    });

    const menuItem = wrapperWithProps.find('.icon');

    await menuItem.trigger('click');

    expect(menuItem.exists()).toBeTruthy();
    expect(wrapperWithProps.emitted('toggle'));
    
  });
});