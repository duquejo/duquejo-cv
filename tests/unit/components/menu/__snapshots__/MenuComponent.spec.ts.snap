// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ MenuComponent.vue tests > Should match the snapshot with the defaults 1`] = `
"<nav class="lg:flex lg:flex-col lg:items-center lg:justify-between hidden lg:my-4 relative">
  <lang-select-component-stub id="menu-language-selector"></lang-select-component-stub>
  <!--v-if--><i class="icon cursor-default" title="general.menu.events_loading_description">
    <v-icon name="fa-spinner" scale="1" fill="gray" animation="spin"></v-icon>
  </i>
</nav>"
`;
